devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java

```diff
@@ -0,0 +1,188 @@
+
+/**
+ * Enforces that for each direct "runtime" dependency the current project also defines a direct minimal "*-deployment"
+ * dependency to produce a consistent build order.
+ */
+public class RequiresMinimalDeploymentDependency implements EnforcerRule2 {
+
+    private static final String GROUP_ID_PREFIX = "io.quarkus";
+    private static final String DEPLOYMENT_ARTIFACT_ID_SUFFIX = "-deployment";
+
+    private static final String EXT_PROPERTIES_PATH = "META-INF/quarkus-extension.properties";
+
+    private static final String REQ_TYPE = "pom";
+    private static final String REQ_SCOPE = "test";
+
+    private static final String DEP_TEMPLATE = "        <dependency>\n"
+            + "            <groupId>%s</groupId>\n"
+            + "            <artifactId>%s</artifactId>\n"
+            + "            <version>${project.version}</version>\n"
+            + "            <type>" + REQ_TYPE + "</type>\n"
+            + "            <scope>" + REQ_SCOPE + "</scope>\n"
+            + "            <exclusions>\n"
+            + "                <exclusion>\n"
+            + "                    <groupId>*</groupId>\n"
+            + "                    <artifactId>*</artifactId>\n"
+            + "                </exclusion>\n"
+            + "            </exclusions>\n"
+            + "        </dependency>";
+
+    private static final Map<String, Optional<String>> DEPLOYMENT_GAV_CACHE = new ConcurrentHashMap<>();
+
+    private Log logger;
+
+    private EnforcerLevel level = EnforcerLevel.ERROR;
+
+    @Override
+    public EnforcerLevel getLevel() {
+        return level;
+    }
+
+    public void setLevel(EnforcerLevel level) {
+        this.level = level;
+    }
+
+    @Override
+    public void execute(EnforcerRuleHelper helper) throws EnforcerRuleException {
+        logger = helper.getLog();
+        MavenProject project;
+        try {
+            project = (MavenProject) helper.evaluate("${project}");
+        } catch (ExpressionEvaluationException e) {
+            throw new IllegalStateException("Failed to get project from EnforcerRuleHelper", e);
+        }
+
+        // general notes:
+        // - "artifacts" are needed to retrieve the actual jar files
+        // - "dependencies" are needed to limit the scope to only the direct dependencies of the current module
+        //   and to check for the proper scope, type and exclusions
+        // - parents are not gathered here since it is expected that the rule is also active for those parents
+        //   or those parents are not relevant
+
+        Map<String, Artifact> nonDeploymentArtifactsByGAV = project.getArtifacts().stream()
+                .filter(artifact -> "jar".equals(artifact.getType()))
+                .filter(artifact -> artifact.getGroupId().startsWith(GROUP_ID_PREFIX))
+                .filter(artifact -> !artifact.getArtifactId().endsWith(DEPLOYMENT_ARTIFACT_ID_SUFFIX))
+                .collect(Collectors.toMap(this::buildGAVKey, a -> a));
```

Comment:
One thing to keep in mind for the future is other extension projects (out there in the community) may want to benefit from this too. In which case GROUP_ID_PREFIX should be either configurable to simply removed in favor of a different approach. E.g. these artifacts are supposed to be resolved at this point, aren't they? We could simply check for the presence of the extension descriptor and also filter based on the resolved path (to collect only those that are found in the current workspace).

Reply:
> E.g. these artifacts are supposed to be resolved at this point, aren't they?

Yes!

> In which case GROUP_ID_PREFIX should be either configurable

Rules can have parameters so this would be pretty simple, yes.

> We could simply check for the presence of the extension descriptor

If I understood you correctly, this would mean we would "open" each jar artifact to check whether it contains a descriptor file. This will come at a price, but maybe it is just a few additional milliseconds in the end.

> also filter based on the resolved path (to collect only those that are found in the current workspace)

Certainly possible but I see problems when building only a subtree of the repo (reactor does not contain all modules in this case).

Reply:
I don't think checking for the presence of the descriptor will be a bottle neck. We are actually doing this check everywhere where we are analyzing the classpath.

As to the subtree, you could use Quarkus bootstrap MavenArtifactResolver instead. There we discover the workspace and it'll work consistently for every build phase and won't depend on the module which was targeted by the mvn cli. I've done this trick in other mojos.

Reply:
> As to the subtree, you could use Quarkus bootstrap MavenArtifactResolver instead

I actually thought about that when I hit that artifact resolve problem but I decided to keep it KISS until we really, really need it.
In the current form, the rule is really "independent".